File: Prefix Sum/PivotIndex.java
import java.util.stream.IntStream;

public class PivotIndex {
  public int pivotIndex(int[] nums) {
    int totalSum = IntStream.of(nums).sum();
    int leftSum = 0;
    for (int i = 0; i < nums.length; i++) {
      int rightSum = totalSum - nums[i] - leftSum;
      if (leftSum == rightSum)
        return i;
      leftSum += nums[i];
    }
    return -1;
  }
}

// Time Complexity: O(n)
// Space Complexity: O(1)

File: Prefix Sum/pivot_index.py
from typing import List

class PivotIndex:
  def pivotIndex(self, nums: List[int]) -> int:
    totalSum = sum(nums)
    leftSum = 0
    for i in range(len(nums)):
      rightSum = totalSum - nums[i] - leftSum
      if leftSum == rightSum: 
        return i
      leftSum += nums[i]

    return -1

# Time complexity: O(n) where n is the length of the input array.
# Space complexity: O(1) as we are using only a constant amount of extra space.

print(PivotIndex().pivotIndex([1, 7, 3, 6, 5, 6]))  # Output: 3